src/jvm/main/org/jetbrains/kotlinx/lincheck/paramgen/ExpandingRangeIntGenerator.kt

```diff
@@ -0,0 +1,158 @@
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>
+ */
+
+package org.jetbrains.kotlinx.lincheck.paramgen
+
+import java.util.*
+
+
+/**
+ * Extensible range-based random generator.
+ *
+ * @param random   random for this generator
+ * @param startInclusive    initial range left bound (inclusive)
+ * @param endInclusive      initial range right bound (inclusive)
+ * @param minStartInclusive minimal range left bound (inclusive)
+ * @param maxEndInclusive   maximal range right bound (inclusive)
+ */
+class ExpandingRangeIntGenerator(
+    private val random: Random,
+    private val startInclusive: Int,
+    private val endInclusive: Int,
+    private val minStartInclusive: Int,
+    private val maxEndInclusive: Int
+) {
+
+    /**
+     * Current left (inclusive) bound of current range
+     */
+    private var currentStartInclusive: Int
+
+    /**
+     * Current right (inclusive) bound of current range
+     */
+    private var currentEndInclusive: Int
+
+    private var nextExpansionDirection = NextExpansionDirection.UP
+
+
+    init {
+        currentStartInclusive = startInclusive
+        currentEndInclusive = endInclusive
+    }
+
+    /**
+     * Reset bounds of this expanding range
+     */
+    fun restart() {
+        currentStartInclusive = startInclusive
+        currentEndInclusive = endInclusive
+    }
+
+    /**
+     * Generates next random number.
+     * With 50% percent probability expands current range if can and returns an expanded bound as a value.
+     * Otherwise, returns a random value from the current range
+     *
+     * @return random value from current range or moved bound
+     */
+    fun nextInt(): Int {
+        checkRangeExpansionAbility()
+        if (nextExpansionDirection == NextExpansionDirection.DISABLED || !random.nextBoolean()) {
+            return generateFromRandomRange(currentStartInclusive, currentEndInclusive)
+        }
+
+        return if (nextExpansionDirection == NextExpansionDirection.DOWN) {
+            nextExpansionDirection = NextExpansionDirection.UP
+            --currentStartInclusive
+        } else {
+            nextExpansionDirection = NextExpansionDirection.DOWN
+            ++currentEndInclusive
+        }
+    }
+
+    private fun checkRangeExpansionAbility() {
+        if (currentStartInclusive == minStartInclusive && currentEndInclusive == maxEndInclusive) {
+            nextExpansionDirection = NextExpansionDirection.DISABLED
+            return
+        }
+        if (currentStartInclusive == minStartInclusive) {
+            nextExpansionDirection = NextExpansionDirection.UP
+            return
+        }
+        if (currentEndInclusive == maxEndInclusive) {
+            nextExpansionDirection = NextExpansionDirection.DOWN
+        }
+    }
+
+    private enum class NextExpansionDirection {
```

Comment:
Please explain somewhere that you first try to generate a new value on the left, then on the right, and so on.